### ISSUE TITLE
[Question] How to get the metadata of every single package on NuGet.org

### ISSUE THREAD

**Issue by jamesqo** [NuGet/Home#6393]
Hi, I want to read the metadata of all 100k+ packages listed on NuGet.org into a giant CSV file so I can use data processing on it. What's the code for this? So far, this is what I have:

```cs
        async static Task Main()
        {
            Logger logger = new Logger();
            IEnumerable<INuGetResourceProvider> providers = new List<INuGetResourceProvider>();
            providers = providers.Concat(Repository.Provider.GetCoreV3().Select(l => l.Value));  // Add v3 API support
            var packageSource = new PackageSource("https://api.nuget.org/v3/index.json");
            SourceRepository sourceRepository = new SourceRepository(packageSource, providers);
            PackageMetadataResource packageMetadataResource = await sourceRepository.GetResourceAsync<PackageMetadataResource>();
            IEnumerable<IPackageSearchMetadata> searchMetadata = await packageMetadataResource.GetMetadataAsync(
                packageId: "",
                includePrerelease: true,
                includeUnlisted: true,
                log: logger,
                token: CancellationToken.None);
            Console.WriteLine($"Count: {searchMetadata.Count()}");
        }
```

This seems to be failing at the `GetMetadataAsync` call when I pass in an empty query.

**Comment by joelverhagen** [comment]
Hey @jamesqo, thanks for taking an interest in this. The code you have right now is using the `PackageMetadataResource` which is useful when you want to find in-depth information about a single package ID. That is, this works when the package ID is known in advance.

What you're asking to do is to enumerate all packages on NuGet -- to discover all package IDs (and presumably versions as well). This is an advanced scenario that is not supported directly by the client today. Since this is a more advanced scenario, it's probably worth becoming at least a little familiar with the [NuGet HTTP API](https://docs.microsoft.com/en-us/nuget/api/overview) which can enable all sorts of analytics and creative applications which are out of scope for the official NuGet client (e.g. nuget.exe). 

I recently wrote a guide to help people discover all packages on NuGet.org using the part of the HTTP API called the [catalog](https://docs.microsoft.com/en-us/nuget/api/catalog-resource). [At the end of the guide](https://docs.microsoft.com/en-us/nuget/guides/api/query-for-all-published-packages#catalog-sdk), I provide some code and even a [package on MyGet](https://dotnet.myget.org/feed/nuget-build/package/nuget/NuGet.Protocol.Catalog) to allow you to easily enumerate packages. The sample code consuming this package is [on GitHub](https://github.com/NuGet/Samples/tree/master/CatalogReaderExample/NuGet.Protocol.Catalog.Sample). If you read this guide, you should be able learn how to enumerate all package IDs and versions on NuGet.org.

If all you need are package IDs and versions, the catalog has everything you care about. Depending on what package metadata you need, the catalog may have everything you need in the [catalog leaves](https://docs.microsoft.com/en-us/nuget/api/catalog-resource#catalog-leaf). However it should be noted that it possible to find deleted or old metadata in the catalog so you can avoid this complexity by simply discovering package IDs and versions from the [catalog pages](https://docs.microsoft.com/en-us/nuget/api/catalog-resource#catalog-page) and going to the package metadata resource find the latest metadata and whether the package is actually deleted as of now. You can do this using the code you posted above or hit the [package metadata API](https://docs.microsoft.com/en-us/nuget/api/registration-base-url-resource) directly with custom code. I would recommend the former so you don't have to write your own JSON mapping to helpful .NET models.

**tl;dr**: use [the "catalog" guide](https://docs.microsoft.com/en-us/nuget/guides/api/query-for-all-published-packages) to discover package IDs and your code above to get up-to-date information about all versions of each ID.

**Comment by joelverhagen** [comment]
@jamesqo, please feel free to re-open this issue if you have any follow-up questions.

**Comment by jamesqo** [comment]
@joelverhagen Thanks for your help, I've managed to get my app working using the catalog API. I have another question about it here: #6435
